pkg/rtc/mediatrack.go

```diff
@@ -652,3 +667,109 @@ func (t *MediaTrack) DebugInfo() map[string]interface{} {
 func (t *MediaTrack) Receiver() sfu.TrackReceiver {
 	return t.receiver
 }
+
+func (t *MediaTrack) OnSubscribedMaxQualityChange(f func(trackSid string, subscribedQualities []*livekit.SubscribedQuality) error) {
+	t.onSubscribedMaxQualityChange = f
+}
+
+func (t *MediaTrack) NotifySubscriberMute(subscriberID string) {
+	if !t.IsSimulcast() {
+		return
+	}
+
+	t.maxQualityLock.Lock()
+	_, ok := t.maxSubscriberQuality[subscriberID]
+	if !ok {
+		t.maxQualityLock.Unlock()
+		return
+	}
+
+	delete(t.maxSubscriberQuality, subscriberID)
+	t.maxQualityLock.Unlock()
+
+	t.updateQualityChange()
+}
+
+func (t *MediaTrack) NotifySubscriberMaxQuality(subscriberID string, quality livekit.VideoQuality) {
+	if !t.IsSimulcast() {
+		return
+	}
+
+	t.maxQualityLock.Lock()
+	maxQuality, ok := t.maxSubscriberQuality[subscriberID]
+	if ok && maxQuality == quality {
+		t.maxQualityLock.Unlock()
+		return
+	}
+
+	t.maxSubscriberQuality[subscriberID] = quality
+	t.maxQualityLock.Unlock()
+
+	t.updateQualityChange()
+}
+
+func (t *MediaTrack) updateQualityChange() {
+	if t.IsMuted() {
+		return
+	}
+
+	var subscribedQualities []*livekit.SubscribedQuality
+
+	t.maxQualityLock.Lock()
+	allSubscribersMuted := false
+	maxSubscribedQuality := livekit.VideoQuality_LOW
+	if len(t.maxSubscriberQuality) == 0 {
+		allSubscribersMuted = true
+	} else {
+		for _, maxQuality := range t.maxSubscriberQuality {
+			switch maxSubscribedQuality {
```

Comment:
I think we are trying to get the max quality from the values in the map? how about just doing the following?

```
for _, subQuality := range t.maxSubscriberQuality {
    if subQuality > maxSubscribedQuality {
        maxSubscribedQuality = subQuality
    }
}
```

Reply:
Oh yeah, they are just integers in the end. Will change. Because it is `enum`, my natural tendency was to think of them as not integers :-)

Reply:
Addressed in this commit - https://github.com/livekit/livekit-server/pull/270/commits/70d443dbe503dd329d5dbe38320f5801b406f3a1